controllers/openstackcluster_controller.go

```diff
@@ -126,20 +126,22 @@ func (r *OpenStackClusterReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		if openStackCluster.Status.Bastion == nil {
 			openStackCluster.Status.Bastion = &infrav1.BastionStatus{}
 		}
-		changed, err := compute.ResolveReferencedMachineResources(scope, openStackCluster, &openStackCluster.Spec.Bastion.Instance, &openStackCluster.Status.Bastion.ReferencedResources)
-		if err != nil {
-			return reconcile.Result{}, err
-		}
-		if changed {
+		if openStackCluster.Status.Bastion.ReferencedResources == nil {
+			openStackCluster.Status.Bastion.ReferencedResources = &infrav1.ReferencedMachineResources{}
+			err := compute.ResolveReferencedMachineResources(scope, openStackCluster, &openStackCluster.Spec.Bastion.Instance, openStackCluster.Status.Bastion.ReferencedResources)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
 			// If the referenced resources have changed, we need to update the OpenStackCluster status now.
 			return reconcile.Result{}, nil
 		}
 
-		changed, err = compute.ResolveDependentBastionResources(scope, openStackCluster, bastionName(cluster.Name))
-		if err != nil {
-			return reconcile.Result{}, err
-		}
-		if changed {
+		if openStackCluster.Status.Bastion.DependentResources == nil {
```

Comment:
DependentResources is different to ReferencedResources because it's not immutable: we can successively create or adopt dependent resources over multiple reconciles. We must always do adoption here.

Reply:
but how would you know it has changed and you must leave the reconcile now to write the object?
If you remove the if, we're in infinite loop where we'll always adopt...